src/editor/BpmnEditor.tsx

```diff
@@ -21,394 +21,422 @@ import { createUIUpdateRequiredEvent } from "../events/modeler/UIUpdateRequiredE
+        onEvent(
+          createNotificationEvent(
+            "Importing element templates failed. Check console for details.",
+            "error",
+          ),
+        );
+        // eslint-disable-next-line no-console
+        console.error("Importing element templates failed.", data);
+      }
+
+      /**
+       * If the event should trigger an UI update required event, do it.
+       */
+      if (event && UI_UPDATE_REQUIRED_EVENTS.indexOf(event) !== -1) {
+        onEvent(createUIUpdateRequiredEvent(active));
+      }
+
+      /**
+       * If the event should trigger a content saved event, do it.
+       */
+      if (event && CONTENT_SAVED_EVENT.indexOf(event) !== -1 && ref.current) {
+        try {
+          const saved = await ref.current.save();
+          onEvent(createContentSavedEvent(saved.xml, saved.svg, "diagram.changed"));
+        } catch (e) {
+          // eslint-disable-next-line no-console
+          console.warn("Could not save document", e);
         }
+      }
     },
-    hidden: {
-        display: "none"
-    },
-    modelerOnly: {
-        height: "100%"
+    [active, onEvent],
+  );
+
+  /**
+   * Instantiates the modeler and properties panel. Only happens once on mount.
+   */
+  useEffect(() => {
+    console.log("Instantiating BPMN Modeler...");
+    const modeler = new CustomBpmnJsModeler({
+      container: modelerOptions?.containerId || "#bpmnview",
+      propertiesPanel: propertiesPanelOptions?.hidden
+        ? undefined
+        : propertiesPanelOptions?.containerId || "#bpmnprop",
+      bpmnJsOptions: bpmnJsOptions,
+    });
+
+    ref.current = modeler;
+    if (modelerOptions?.refs) {
+      modelerOptions.refs.forEach((r) => {
+        // eslint-disable-next-line no-param-reassign
+        r.current = modeler;
+      });
     }
-}));
 
-const BpmnEditor: React.FC<BpmnEditorProps> = props => {
-    const classes = useStyles();
-
-    const {
-        active,
-        xml,
-        onEvent,
-        className,
-        bpmnJsOptions,
-        modelerOptions,
-        propertiesPanelOptions
-    } = props;
-
-    const [initializeCount, setInitializeCount] = useState(0);
-    const ref = useRef<CustomBpmnJsModeler | undefined>(undefined);
-
-    const handleEvent = useCallback(async (event: string, data: any) => {
-        // TODO: Should bpmn-js events only be forwarded if the editor is currently active?
-        onEvent(createBpmnIoEvent(event, data));
-
-        if (!active) {
+    setInitializeCount((count) => count + 1);
+
+    return () => {
+      console.log("Destroying BPMN Modeler...");
+      modeler.unregisterGlobalEventListener(handleEvent);
+      modeler.destroy();
+      ref.current = undefined;
+      if (modelerOptions?.refs) {
+        modelerOptions.refs.forEach((r) => {
+          // eslint-disable-next-line no-param-reassign
+          r.current = undefined;
+        });
+      }
+    };
+  }, [
+    bpmnJsOptions,
+    // modelerOptions,
```

Comment:
Well yes I get a warning. But I removed it because if I change something inside the diagram the ref object changes and causes the modeler to re-render.